### PR TITLE
Add `PostgresSchema` to manage Postgres schemas with

### PR DESCRIPTION
**Pull request by Photonios**

### CODE REVIEW

File: tests/test_schema.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        cursor.execute("SELECT * FROM test.bla")
+
+        assert cursor.fetchone() == ("hello",)
+
+    # Should refuse to delete since we added a table to the schema
+    with pytest.raises(InternalError) as exc_info:
+        schema = PostgresSchema.delete_and_create(schema.name)
+
+    pg_error = extract_postgres_error(exc_info.value)
+    assert pg_error.pgcode == errorcodes.DEPENDENT_OBJECTS_STILL_EXIST
```
Comment (sewi-cpan) [inline tests/test_schema.py:77]: A test for the table would be nice here. Unlikely to get an error while the schema is still being deleted but the test is cheap.

File: psqlextra/schema.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    def cursor(self) -> CursorWrapper:
+        schema = self._self_schema
+
+        with self.__wrapped__.cursor() as cursor:
+            quoted_name = self.ops.quote_name(schema.name)
+            cursor.execute(f"SET search_path = {quoted_name}")
+            try:
+                yield cursor
+            finally:
+                cursor.execute("SET search_path TO DEFAULT")
```
[the remark this answers is not in the document]
Comment (Photonios) [reply psqlextra/schema.py:37]: It won't, you're supposed to use this with a direct database connection. I will leave a warning in the documentation.

File: psqlextra/schema.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        Arguments:
+            prefix:
+                Name to prefix the final name with. The name plus
+                prefix cannot be longer than 63 characters.
+
+            using:
+                Name of the database connection to use.
+        """
+
+        name_suffix = timezone.now().strftime("%Y%m%d%H%m%s")
```
[the remark this answers is not in the document]
Comment (Photonios) [reply psqlextra/schema.py]: I was aware of this possible race condition and I didn't want to use in the scenarios you described. My use case for this method is to create a schema with a backup. Having a date-time based name is good for that as it quickly allows you to determine which is older and which is newer.

The name is a bit misleading as it creates the suggestion that it's truly random. To solve this I have done the following:

1. Rename `create_random` to `create_time_based` and documented that it's precise to the second and that it's subject to race conditions.

2. Create a new `create_random` that uses `os.urandom(4)`, which is backed by a large amount of entropy.

> Does a length check make sense if the prefix plus the random part are longer than 63 chars? Or is this unlikely to happen and people need to debug what's going on if the create() exception appears?

I've added this.

File: tests/test_schema.py (hunk 1 above)
Comment (Photonios) [reply tests/test_schema.py:103]: Added

File: psqlextra/schema.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        a problem that requires careful handling. Pretending everything
+        is ok might cause the caller to overwrite data, thinking it got
+        a empty schema.
+
+        Arguments:
+            name:
+                The name to give to the new schema (max 63 characters).
+
+            using:
+                Name of the database connection to use.
```
[the remark this answers is not in the document]
Comment (Photonios) [reply psqlextra/schema.py]: I've updated the doc string to be the same as in the other PR.

However, I don't think it is very necessary to specify this for all parameters. The generated documentation clearly shows parameters with a default:

Screenshot 2023-04-06 at 11 17 19

File: psqlextra/schema.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Photonios) [reply psqlextra/schema.py]: I changed my mind in the end and removed the entire feature. It's too tricky to get right and it's too easy for users to shoot themselves in the foot with.

File: psqlextra/schema.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+            )
+
+        with connections[using].schema_editor() as schema_editor:
+            schema_editor.delete_schema(self.name, cascade=cascade)
+
+    @classmethod
+    def _verify_generated_name_length(cls, prefix: str, suffix: str) -> None:
+        max_prefix_length = cls.NAME_MAX_LENGTH - len(suffix)
+
+        if len(prefix) > max_prefix_length:
```
Comment (sewi-cpan) [inline psqlextra/schema.py:166]: Assumptions:
1. Prefix len 50
2. Suffix len 13
`max_prefix_length` would be `63 - 13` = `50`
`len(prefix)` = `50` is not `> max_prefix_length` `50`.
The calling function does `{prefix}_{suffix}` which results in 50 **+ 1** + 13 = 64 chars.

Either remove the seperator from the creation functions or consider it here while calculating.

File: docs/source/schemas.rst [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+   for psqlextra.schema import PostgresSchema
+
+   schema = PostgresSchema("myschema")
+
+   with schema.connection.cursor() as cursor:
+       cursor.execute("SELECT * FROM tablethatexistsinmyschema")
+
+
+Checking if a schema exists
+---------------------------
```
[the remark this answers is not in the document]
Comment (Photonios) [reply docs/source/schemas.rst:29]: I don't think the dark red indicates that it was removed. The `-` are all there. See the raw version:

https://raw.githubusercontent.com/SectorLabs/django-postgres-extra/schema-management/docs/source/schemas.rst

File: psqlextra/schema.py (hunk 5 above)
Comment (Photonios) [reply psqlextra/schema.py]: Oops, forgot about the separator. Taken into account now.

**Review by sewi-cpan — APPROVED** [review]

**Pull request merged**